buildSrc/src/main/kotlin/codegen/math/GenerateMath.kt

```diff
@@ -0,0 +1,249 @@
+package codegen.math
+
+import com.squareup.kotlinpoet.*
+import com.squareup.kotlinpoet.KModifier.*
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+
+@Suppress("UnstableApiUsage")
+open class GenerateMath : DefaultTask() {
+	@OutputDirectory
+	val outputDir = project.objects.directoryProperty()
+
+	@get:OutputDirectory
+	val commonDir = outputDir.file("common")
+
+
+	private val packageName = "com.kgl.math"
+
+	private val typeList = listOf(
+		BYTE to "0",
+		SHORT to "0",
+		INT to "0",
+		LONG to "0L",
+		U_BYTE to "0U",
+		U_SHORT to "0U",
+		U_INT to "0U",
+		U_LONG to "0UL",
+		FLOAT to "0f",
+		DOUBLE to "0.0"
+	)
+
+
+	@TaskAction
+	fun generate() {
+		vectorTypes()
+	}
+
+	private val allComponents = listOf(1 to "x", 2 to "y", 3 to "z", 4 to "w")
+	private val arithmetics = listOf("plus" to "+", "minus" to "-", "times" to "*", "div" to "/")
+
+	private fun vectorTypes() {
+		for (componentCount in 2..4) {
+			val components = allComponents.take(componentCount)
+
+			buildFile(packageName, "Vector$componentCount") {
+				import("kotlin.math", "sqrt")
+
+				typeList.forEach { (type, zero) ->
+					val vectorType = ClassName(packageName, "${type.simpleName}Vector$componentCount")
+					val mutableVectorType = ClassName(packageName, "Mutable${vectorType.simpleName}")
+
+					buildClass(vectorType) {
+						modifiers(SEALED)
+
+						components.forEach { (_, name) ->
+							property(name, type, ABSTRACT)
+						}
+
+						function("get") {
+							modifiers(OPERATOR)
+							parameter("index", INT)
+							controlFlow("when (index)") {
+								components.forEach { (i, name) ->
+									statement("${i - 1} -> $name")
+								}
+								statement("else -> throw IndexOutOfBoundsException()")
+							}
+						}
+
+						property("length", type) {
+							getter {
+								val code = when (type) {
+									BYTE, SHORT, INT, LONG ->
+										"return sqrt((this dot this).toDouble()).to${type.simpleName}()"
+									U_BYTE, U_SHORT, U_INT, U_LONG ->
+										"return sqrt((this dot this).toDouble()).to${type.simpleName.drop(1)}().to${type.simpleName}()"
+									else ->
+										"return sqrt(this dot this)"
```

Comment:
Should be `sqrt(lengthSquared)`.

Reply:
I was intending to save a function call (of the `lengthSquared` getter). Should I not?

Reply:
Hmm. I trust the JVM to inline it but I'm not sure about LLVM. We can leave it until we start bench-marking.